### PR TITLE
CommandLineOptions tests POC

### PR DESCRIPTION
**Pull request by rvanheest**
related to EASY-1049 - *Upgrade Scallop*

@DANS-KNAW/easy 

While doing the first batch of PRs for EASY-1049, I felt quite insecure about some of the changes I made. After all, there are no automated tests to check whether the CLIs are still working as expected. The way to test this currently is building, deploying and testing on DEASY. This is however quite a hassle and often involves too much work to be bothered with. So the *unofficial official way* to test this is simply **not to test it**.

My fears about breaking some behavior unexpectedly turned out to come true in a couple of projects, but thanks to @lindareijnhoudt they got caught early on and before any PRs were merged!

To prevent these things from happening in the future, I sat down and created a simple Proof of Concept (POC) of how this could be tested using our ScalaTest framework. The results of this POC can be found in this PR. **It is not necessarily something to be merged immediately, but more as a discussion starter to see what we can do with this!**

***Side note on this PR:** the changes in `pom.xml` and `CommandLineOptions:44` belong to PR #13, which was not merged yet at the time of writing this, but was still necessary for the current PR.*

A couple of conclusions can be drawn from this experiment:
1. Scallop does not particularly like it when you provide it with bad, invalid or unknown arguments and will literally call `System.exit` on them! Therefore in the tests that are supposed to fail, we require to **override the `errorMessageHandler` lambda variable** such that we can **catch the error message** and **prevent the testing framework from terminating** in this way.
2. Because of this we cannot call `verify()` at the end of the argument definition. If we did, the override of the `errorMessageHandler` would only occur after the arguments were already verified and after `System.exit` had already been called.
    * **Who should call `verify()` instead?** - `verify()` should be called by the class that also creates the `CommandLineOptions` class, most preferably immediately after its creation and at least before retrieving the actual values from the class itself.
    * **What happens when we forget to call `verify()`?** - Of course it is quite a risk to let the user of the `CommandLineOptions` class call `verify()`. However, Scallop will make sure that the actual values can only be retrieved **after `verify()` has been called**! If this line is called too late or not at all, it will cause an appropriate `IncompleteBuildException`, saying: 

        > It seems you tried to get option value before you constructed all options (maybe you forgot to call .verify method?). Please, move all extraction of values to after 'verify' method in ScallopConf.

Let me know if adding this would in general have any additional value to the CLI driven projects and whether we should include this in more projects!

### CODE REVIEW

**Comment by rvanheest** [comment]
@jo-pol came up with the very good idea (see https://github.com/rvanheest-DANS-KNAW/easy-license-creator/pull/2) to make the testing code less repetative and more clear to read. After some discussion we came up with the `CLISpec` trait, which contains all CLI creation code. Tests are even more concise this way. For now we suggest copy pasting this trait to the repositories where we need to use it.

**Comment by rvanheest** [comment]
_**TLDR:** indeed they somehow test the Scallop library, but they already helped me find a small bug in one of the projects and therefore seem pretty valuable to me._

Mmm, I see what you mean. And in a sense there's a bit of truth in that, in the sense that expressions like `opts.isSample() shouldBe false` do partially test the functionality of the Scallop library.

However, we also test our own code by seeing which combinations of command line arguments are (in)valid. For this POC I choose an easy example project, but other projects have dependencies between the parameters or extra conditions on the parameters (like a `File` existing, a directory containing at least one other directory, etc.), which IMO need to be properly tested. In that case we are not testing the functionality of the library itself, but rather the constraints we have laid on our CLI.

In a sense we do that in this project as well. If you look at the test you see three 'succeeding' tests, where we actually say: "*It is OK to either do or do not include an `-s` or `--sample` flag and here is the output we expect whenever this flag is or isn't present.*" Although it is a fairly small example in this case, I think it seems very valuable to have tests like these.

Finally, as I told you I wasn't sure about all refactorings in the Scallop issue, so I applied this test suite to the `easy-validate-aip` project and found a bug in my refactoring, which I got fixed before committing/merging. Because of this POC I did not include those tests yet, but I still have them on a local branch on my computer. I think they really helped me to keep the CLI as intended and not introduce bugs in the final commit!

**Comment by janvanmansum** [comment]
I don't object to unit tests calling libraries per se. It is true that in doing so you are in fact partly testing that the library does its job. Actually, you implicitly *assume* that it does its job. If that assumption fails you find a bug in that library - not your original goal, but a useful by-catch anyway.

That is all well and good, but the focus of your tests should of course be on testing our *own* code. (After all, if you don't trust the libraries you are using, you should probably use different ones, or improve them, and thereby *making* them your own code.)

To be frank, the fact that you encountered a bug while writing a test, does not necessarily tell me that the test is good. Whenever you exercise a piece of code, you will find bugs. 

So, how *do* we evaluate a unit test? I think the following questions can be useful:

1. What feature or property of our code does the test verify?
2. Wat part of our code does the test exercise?

Basically, the answer to question 1 would name a part of the unit interface contract. So, if we test a function, we should be able to specify what about that function's contract is being tested. If we can't or don't I think the unit test is not good or maybe even bad. So, tests like: "just trying to see if I got the same results as when I used it before", are not proper unit tests in my view.

Number 2 is more relevant for the current discussion. Basically, it is about what lines of code get touched. Obviously if you implement a complex search algorithm by yourself in, say, Java-code a lot more testing is needed than when your search function just calls a search library, even though the feature and interface may be exactly the same in both cases.

Let me try to answer both questions for the POC. Maybe that is a little unfair, as a POC may be more trivial than usual. However, it is a place to start.

1. What features of the code gets tested? This should be in the description of the tests. So I am finding the following (in my own words), in the order in which the tests are defined in   `nl.knaw.dans.easy.license.app.CommandLineOptionsSpec`:
    
    - (a) Short notation for option gets parsed correctly
    - (b) Long notation for option gets parsed correctly
    - (c) Trailing param gets parsed correctly (or: "options are not required")
    - (d) Missing required trailing arg leads to error
    - (e) Invalid option lead to error

   These are all actual features of the command line interface, so in that respect they are sound unit tests in my view.
2. What part of the code gets exercised? 

    - (a) I'd say we are testing that we specified the named param `short = 's'` to the call to Scallop's `opt`
    - (b) Idem for `name = "sample"`
    - (c) We're verifying that we declared a trailing arg (the 2 lines starting with `val datasetID`
    - (d) We're testing that both declared trailing arguments are required which is nowhere explicitly declared in our code.

    So, the way I see it, we are not testing any logic in our own code. Had we implemented the command line parsing ourselves, the tests would indeed be very appropriate, but now the code is already very declarative and, to the extent that we are not simply testing that Scallop is doing its    job, the tests declare the exact same things as the code itself.
    
Okay, this is becoming a rather long-winded answer, but I am partly just thinking aloud. Maybe there are cases in which we have more logic in our own code that augments the behavior of Scallop. Then this kind of test could indeed be useful. In fact, the check that @jo-pol built in in a couple of modules, which checks that the arguments docs in README.md is in sync with Scallop is a good example.

**Comment by rvanheest** [comment]
I fully agree with these two metrics for useful unit tests in general. However, I do not completely agree with your metaphor about a search algorithm. Yes, you require lots of tests to guarantee its correctness (maybe even a more property based testing approach) when you implement it yourself. However, this search algorithm is most definitely (part of) an interface that you use in other places. These other places do indeed not require to test the correctness of the algorithm itself, but rather the logic that surrounds the call to this algorithm: did it receive the correct input, is the output processed correctly, etc. *These tests* are equal in both the case that you implemented the algorithm yourself and the case that you use an external library. (Typically keywords like *mocking* are involved in these kinds of tests, but I leave those out for now.)

In my view, the tests in this POC are equal to this second type of tests: not checking the algorithm itself but rather testing the *use* of the algorithm. What we specify in `CommandLineOptions.scala` is the *use* of the Scallop library and with that declare a contract that the user of our CLI has to obey. What we test in the `spec` is the correct declaration of this contract. We're not just testing that Scallop works correctly (I've seen their tests and they already do that enough), we're testing the correctness of our own declarations.

I further agree with your analysis of "*What features of the code are being tested*": I think your analysis of what these tests do is correct. However, I want to add and repeat that I choose this project on purpose as it does not have additional constraints on the declared CLI parameters. In the case that (for example) we wanted the `outputFile` argument to represent an explicitly *non-existing file*, we require an extra test for that to check that we declared in our contract (i.e. `CommandLineOptions.scala`) that this is in fact the case.

Regarding the second question, I think this can be answered better like this:
	* (a)/(b)/(c) we test that the contract specifies a boolean flag that is optional
	* (d) we test that the contract specifies two *mandatory* trailing arguments

If we would add the specification for `outputFile` to be a non-existing file* to the contract, we would add an extra test that tests this using both a non-existing file and an existing file, which would succeed and fail respectively. This test would test the 'part of the code' that declares the `outputFile` to be non-existing.

Summarized, these tests are to be seen as a validation of our specification/contract about the CLI; not as additional tests for the functionality of Scallop. Regarding the check by @jo-pol that checks the `README.md`, I think this is a nice addition to the test suite to see whether the contract that is declared in `CommandLineOptions.scala`, is also correctly reflected in the `README.md`. I don't think it can be seen as a correct replacement for these tests, since the contract in `README.md` is just a copy-paste of the CLI's `--help` text. I will add this additional test later today.

Hopefully this makes the intention of these tests somewhat clearer!

**Comment by janvanmansum** [comment]
Your intentions are crystal clear. However, what I was trying to point out is that, since Scallop already has a very declarative interface, it doesn't add anything to test the fact that you called it in certain way. For example the test about the short parameter name, we declare right there in the command line options code that the short option must be 's'. Why test that? Are we testing that we didn't forget to add a declaration? If we are going down that path we'll end up adding a huge amount of tests that only fail because we changed some code and forgot to update the tests.

**Comment by jo-pol** [comment]
Shorts might be implicit or explicit and conflicting, don't remember exactly the scallop behaviour, but I do remember making mistakes in the past. So we might indeed _need_ to verify the short matches the long we think it should match.

**Comment by janvanmansum** [comment]
I am closing this PR because, as explained at length above, I think these particular tests are not testing anything useful. The way of testing the command line is perfectly fine though, so if there is an oportunity to test any command line processing logic of our own then this would be the way to do it. Maybe it is wise to keep a link to the PR to be able to refer back to it.

**Comment by rvanheest** [comment]
Sure, I'll keep the branch around!
